cpp/src/lists/contains.cu

```diff
@@ -44,163 +41,156 @@
+ */
+struct dispatch_index_of {
+  template <typename Type, typename... Args>
+  std::enable_if_t<!is_supported_non_nested_type<Type>(), std::unique_ptr<column>> operator()(
+    Args&&...) const
   {
-    auto output_iterator = thrust::make_zip_iterator(
-      thrust::make_tuple(ret_positions.data<size_type>(), ret_validity.data<bool>()));
-
-    thrust::tabulate(
-      rmm::exec_policy(stream),
-      output_iterator,
-      output_iterator + d_lists.size(),
-      [d_lists, search_key_pair_iter, absent_index = absent_index, find_option] __device__(
-        auto row_index) -> thrust::pair<size_type, bool> {
-        auto [search_key, search_key_is_valid] = search_key_pair_iter[row_index];
-
-        if (search_keys_have_nulls && !search_key_is_valid) { return {absent_index, false}; }
-
-        auto list = cudf::list_device_view(d_lists, row_index);
-        if (list.is_null()) { return {absent_index, false}; }
-
-        auto const position = find_option == duplicate_find_option::FIND_FIRST
-                                ? finder<duplicate_find_option::FIND_FIRST>{}(list, search_key)
-                                : finder<duplicate_find_option::FIND_LAST>{}(list, search_key);
-        return {position, true};
-      });
+    CUDF_FAIL("Unsupported type in `dispatch_index_of` functor.");
   }
 
-  template <typename ElementType, typename SearchKeyType>
-  std::enable_if_t<is_supported<ElementType>::value, std::unique_ptr<column>> operator()(
-    cudf::lists_column_view const& lists,
-    SearchKeyType const& search_key,
+  template <typename Type, typename SearchKeyType>
+  std::enable_if_t<is_supported_non_nested_type<Type>(), std::unique_ptr<column>> operator()(
+    lists_column_view const& lists,
+    SearchKeyType const& search_keys,
+    bool search_keys_have_nulls,
     duplicate_find_option find_option,
     rmm::cuda_stream_view stream,
-    rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource()) const
+    rmm::mr::device_memory_resource* mr) const
   {
-    using namespace cudf;
-    using namespace cudf::detail;
-
     CUDF_EXPECTS(!cudf::is_nested(lists.child().type()),
                  "Nested types not supported in list search operations.");
-    CUDF_EXPECTS(lists.child().type() == search_key.type(),
+    CUDF_EXPECTS(lists.child().type() == search_keys.type(),
                  "Type/Scale of search key does not match list column element type.");
-    CUDF_EXPECTS(search_key.type().id() != type_id::EMPTY, "Type cannot be empty.");
+    CUDF_EXPECTS(search_keys.type().id() != type_id::EMPTY, "Type cannot be empty.");
 
     auto constexpr search_key_is_scalar = std::is_same_v<SearchKeyType, cudf::scalar>;
-
-    if constexpr (search_keys_have_nulls && search_key_is_scalar) {
+    if (search_key_is_scalar && search_keys_have_nulls) {
+      // If the scalar key is invalid/null, the entire output column will be all nulls.
       return make_numeric_column(data_type(type_id::INT32),
```

Comment:
Avoid hard-coding `INT32` when we mean `size_type`. This is the same pattern followed below.
```suggestion
      return make_numeric_column(data_type{cudf::type_to_id<size_type>()},
```